### PR TITLE
Update __init__.py

### PR DESCRIPTION
**Pull request by Shivamgera**
I did [Issue#164](https://github.com/zulip/python-zulip-api/issues/164). Removed RuntimeError and created a custom error called ZulipError. Also, used some more descriptive errors.

### CODE REVIEW

**Comment by roberthoenig** [comment]
@Shivamgera Travis is failing because of this test in test_defaultarguments.py:
```.py
        with self.assertRaises(RuntimeError) as cm:
            zulip.init_from_options(args)
        expanded_test_path = os.path.abspath(os.path.expanduser(test_path))
        self.assertEqual(str(cm.exception), 'api_key or email not specified and '
                         'file {} does not exist'.format(expanded_test_path))
```

This test asserts that a RuntimeError exception is thrown. Since this PR changes the RuntimeErrors to ZulipErrors, it needs to change the test assertion as well.

File: zulip/zulip/__init__.py [hunk 1]
```diff
@@ -267,12 +267,25 @@ def get_default_config_filename():
...
     if (not os.path.exists(config_file) and
             os.path.exists(os.path.join(os.environ["HOME"], ".humbugrc"))):
-        raise RuntimeError("The Zulip API configuration file is now ~/.zuliprc; please run:\n\n"
+        raise ZulipError("The Zulip API configuration file is now ~/.zuliprc; please run:\n\n"
                            "  mv ~/.humbugrc ~/.zuliprc\n")
     return config_file
 
 class ZulipError(Exception):
-    pass
+    def __init__(self,msg):
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: You don't need constructors for a new exception: If it inherits from Exception, the message output is handled automatically.

File: zulip/zulip/__init__.py [hunk 2]
```diff
@@ -362,7 +375,7 @@ def __init__(self, email=None, api_key=None, config_file=None,
             self.tls_verification = False  # type: Union[bool, str]
         elif cert_bundle is not None:
             if not os.path.isfile(cert_bundle):
-                raise RuntimeError("tls bundle '%s' does not exist"
+                raise VerificationError("tls bundle '%s' does not exist"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Should this use "FileNotFoundError"?

File: zulip/zulip/__init__.py [hunk 3]
```diff
@@ -371,15 +384,15 @@ def __init__(self, email=None, api_key=None, config_file=None,
 
         if client_cert is None:
             if client_cert_key is not None:
-                raise RuntimeError("client cert key '%s' specified, but no client cert public part provided"
+                raise ZulipError("client cert key '%s' specified, but no client cert public part provided"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Should this use "FileNotFoundError"?

File: zulip/zulip/__init__.py [hunk 4]
```diff
@@ -371,15 +384,15 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
                                    % (client_cert_key,))
         else:  # we have a client cert
             if not os.path.isfile(client_cert):
-                raise RuntimeError("client cert '%s' does not exist"
+                raise ZulipError("client cert '%s' does not exist"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Should this use "FileNotFoundError"?

File: zulip/zulip/__init__.py [hunk 5]
```diff
@@ -371,15 +384,15 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
                                    % (client_cert,))
             if client_cert_key is not None:
                 if not os.path.isfile(client_cert_key):
-                    raise RuntimeError("client cert key '%s' does not exist"
+                    raise ZulipError("client cert key '%s' does not exist"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Should this use "FileNotFoundError"?

File: zulip/zulip/__init__.py [hunk 6]
```diff
@@ -267,12 +267,25 @@ def get_default_config_filename():
...
+        raise ZulipError("The Zulip API configuration file is now ~/.zuliprc; please run:\n\n"
                            "  mv ~/.humbugrc ~/.zuliprc\n")
     return config_file
 
 class ZulipError(Exception):
-    pass
+    def __init__(self,msg):
+        self.msg = msg
+
+class FileNotExistError(RuntimeError):
```
[the remark this answers is not in the document]
Comment (roberthoenig) [reply zulip/zulip/__init__.py]: It should probably inherit from Exception.

File: zulip/zulip/__init__.py [hunk 7]
```diff
@@ -42,6 +42,7 @@
 import six
 from typing import Any, Callable, Dict, Iterable, IO, List, Mapping, Optional, Text, Tuple, Union
 
+
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Remove this newline.

File: zulip/zulip/__init__.py [hunk 8]
```diff
@@ -267,13 +268,20 @@ def get_default_config_filename():
...
 class ZulipError(Exception):
     pass
 
+class FileNotFoundError(ZulipError):
+    pass
+
+class MissingURLError(ZulipError):
+    pass
+
+
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: Remove one newline.

File: zulip/zulip/__init__.py [hunk 9]
```diff
@@ -362,25 +370,25 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
             self.tls_verification = cert_bundle
         else:
             # Default behavior: verify against system CA certificates
             self.tls_verification = True
 
         if client_cert is None:
             if client_cert_key is not None:
-                raise RuntimeError("client cert key '%s' specified, but no client cert public part provided"
-                                   % (client_cert_key,))
+                raise FileNotFoundError("client cert key '%s' specified, but no client cert public part provided"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: I'm sorry, I overlooked that client_cert and client_cert key can actually be specified as environment vars. Can you change it back to ZulipError for this and the following two exceptions?

**Comment by Shivamgera** [comment]
@roberthoenig I've tried to address the remaining issues. Is it fine now?

File: zulip/zulip/__init__.py [hunk 10]
```diff
@@ -331,11 +337,11 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
                     insecure = False
                 else:
-                    raise RuntimeError("insecure is set to '%s', it must be 'true' or 'false' if it is used in %s"
-                                       % (insecure_setting, config_file))
+                    raise ZulipError("insecure is set to '%s', it must be 'true' or 'false' if it is used in %s"
+                                     % (insecure_setting, config_file))
         elif None in (api_key, email):
-            raise RuntimeError("api_key or email not specified and file %s does not exist"
-                               % (config_file,))
+            raise FileNotFoundError("api_key or email not specified and file %s does not exist"
```
Comment (showell) [inline zulip/zulip/__init__.py]: We should call this `ConfigNotFoundError`, since we're not necessarily looking for a file here.

**Comment by showell** [comment]
@Shivamgera Can you make the first line of the commit message more descriptiive?  You can say something like "Use custom exceptions in API client".

File: zulip/zulip/__init__.py (hunk 10 above)
Comment (Shivamgera) [reply zulip/zulip/__init__.py]: So should we change all FileNotFoundError with ConfigNotFoundError or just this one?

**Comment by Shivamgera** [comment]
@showell  Please review the latest commit.

**Comment by roberthoenig** [comment]
@Shivamgera your commit message body now contains a lot of redundant text. Can you remove it? Additionally,  add dots (.) at the end of each sentence.

**Comment by Shivamgera** [comment]
@roberthoenig I've updated the commit message.

File: zulip/zulip/__init__.py [hunk 11]
```diff
@@ -362,25 +368,25 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
             self.tls_verification = cert_bundle
         else:
             # Default behavior: verify against system CA certificates
             self.tls_verification = True
 
         if client_cert is None:
             if client_cert_key is not None:
-                raise RuntimeError("client cert key '%s' specified, but no client cert public part provided"
-                                   % (client_cert_key,))
+                raise ZulipError("client cert key '%s' specified, but no client cert public part provided"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: s/ZulipError/ConfigNotFoundError/

File: zulip/zulip/__init__.py [hunk 12]
```diff
@@ -362,25 +368,25 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
+                                 % (client_cert_key,))
         else:  # we have a client cert
             if not os.path.isfile(client_cert):
-                raise RuntimeError("client cert '%s' does not exist"
-                                   % (client_cert,))
+                raise ZulipError("client cert '%s' does not exist"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: s/ZulipError/ConfigNotFoundError/

File: zulip/zulip/__init__.py [hunk 13]
```diff
@@ -362,25 +368,25 @@ def __init__(self, email=None, api_key=None, config_file=None,
...
+                                 % (client_cert,))
             if client_cert_key is not None:
                 if not os.path.isfile(client_cert_key):
-                    raise RuntimeError("client cert key '%s' does not exist"
-                                       % (client_cert_key,))
+                    raise ZulipError("client cert key '%s' does not exist"
```
Comment (roberthoenig) [inline zulip/zulip/__init__.py]: s/ZulipError/ConfigNotFoundError/

**Comment by Shivamgera** [comment]
@roberthoenig Done. You're welcome :)

**Comment by roberthoenig** [comment]
@showell I think this is ready.

**Comment by showell** [comment]
@timabbott Can you give this a quick look?  I think it's what we're looking for, but I just want to double check since this affects the core API.

**Comment by timabbott** [comment]
lgtm, merged after cleaning up the commit message.  Thanks @Shivamgera and @roberthoenig for all the work reviewing this!

**Pull request closed without merging**